Add new APIs and fix bugs

Abasz opened the pull request:
This PR includes several improvements of the web ant plus library. It updates certain dependencies as well as include new APIs and several fixes. 

Most importantly it adds new static factory methods to create a USBDriver which essentially bypasses the need to specifically instantiate a GarminStick class instance before a USBDriver can be created. These new methods allow the creation of USBDriver from already paired devices (i.e. no need for selecting from the prompt) and start USBDriver from the USB device selected from the pairing prompt.

It also updates the documentation in the README to reflect the API and corrects some incorrect descriptions

Commit:
Update dependencies and dev stack

Commit:
Fix bug #5

Remove `usbDevice.reset()` call in order to avoid crash as well as add
`await` for the `usbDevice.close()`

Commit:
Enable cancelling readInEndPoint

Add cancellation functionality to the `readInEndPoint` loop to enable
graceful shutdown of the device when its being closed/reset as well as
to be able distinguished an actual device error

Commit:
Add static factory functions to create USBDriver

Enable the creation of a `USBDriver` instance via different static
functions that do not necessarily require prior knowledge of the stick
type (GarminStick2/GarminStick3).

Commit:
Update documentation

8beeeaaat commented on package.json:
```diff
@@ -40,13 +40,12 @@
     "@types/react-dom": "^18.0.8",
     "@types/w3c-web-usb": "^1.0.6",
-    "@typescript-eslint/eslint-plugin": "^5.42.1",
-    "@vitejs/plugin-react": "^2.2.0",
```
What is the reason for removing `@vitejs/plugin-react`?
This plugin should not be removed because it is used in example/vite.config.ts.

![スクリーンショット 2023-10-22 12 07 37](https://github.com/8beeeaaat/web-ant-plus/assets/4495546/f35bb55e-0fcc-46f9-a5a0-e4ba75eebd66)

https://github.com/8beeeaaat/web-ant-plus/blob/7916f5cebdee2f3080e1437b62ca0837181e75ed/example/vite.config.ts#L1-L7

Abasz replied:
Deleted by accident, re adding.

8beeeaaat commented on src/index.ts:
```diff
@@ -1,7 +1,7 @@
 export { GarminStick2 } from './GarminStick2';
 export { GarminStick3 } from './GarminStick3';
-export type { ICancellationToken } from './ICancellationToken';
+export { ICancellationToken } from './ICancellationToken';
```
It should be the implementation `CancellationToken` that exports.
Have you actually run `npm run dev` to check the behavior?

>Uncaught SyntaxError: The requested module '/@fs/Users/UserName/git/web-ant-plus-1/src/ICancellationToken.ts' does not provide an export named 'ICancellationToken' (at index.ts:4:1)

![スクリーンショット 2023-10-22 12 12 28](https://github.com/8beeeaaat/web-ant-plus/assets/4495546/bb476506-6907-4072-9522-6d76ed00b0e0)

Abasz replied:
Fix in next commit.

8beeeaaat commented on tsconfig.json:
```diff
@@ -1,17 +1,25 @@
-  "include": ["./src/**/*.ts"],
+  "include": [
+    "./src/**/*.ts",
+    "src/index.ts"
```
Please I would like to see why you want to include `src/index.ts`.

Abasz replied:
Added by accident. removing

8beeeaaat commented on .editorconfig:
```diff
@@ -0,0 +1,15 @@
+quote_type = single
+indent_style = space
+max_line_length = 80
+insert_final_newline = false
```
I am sorry for the suggestion, but please let `insert_final_newline` be true. 

```suggestion
```

Abasz replied:
Accepted.

8beeeaaat reviewed (changes requested):
Thanks for the PR, sorry I neglected to review your PR! 
I commented because I had a few requests for corrections and some points I wanted to ask for clarification.

Abasz commented:
> "@vitejs/plugin-react": "^2.2.0",

Thanks for taking a look. The issues raised are totally valid and most of them are silly mistakes, I will fix them shortly. Indeed I should have tested if `run dev` works... sorry about that.

8beeeaaat commented on .vscode/settings.json:
```diff
@@ -0,0 +1,12 @@
+  "prettier.trailingComma": "none",
+  "prettier.singleQuote": true,
+  "prettier.jsxSingleQuote": true,
+}
```
please insert final newline :)
other files too.

8beeeaaat reviewed (changes requested):
Thanks for the fix commit!
Finally, I would appreciate it if you could run `npx prettier --write . / && npx eslint --fix` and commit the differences.

When writing or changing a rule, if the developer who implements the change does not apply the fix to the existing file, there will be an extra difference for later development.

Abasz commented:
> Thanks for the fix commit! Finally, I would appreciate it if you could run `npx prettier --write . / && npx eslint --fix` and commit the differences.
> 
> When writing or changing a rule, if the developer who implements the change does not apply the fix to the existing file, there will be an extra difference for later development.

Your point is valid and now I remember that I have tried to do running a full lint before doing the PR. The problem is that when I run eslint on all files there are multiple layers of issues (note that these are not only on my branch but on the original master too):
1. the eslint file is invalid on my machine as it throws this: `Environment key "w3c-web-usb" is unknown`. Not sure about this error
2. removing the w3c-web-usb environment allows eslint to run but then there are almost 250 errors.

Just a side note: actually your quoted command `npx eslint --fix` is a no-op as it needs the list of files to actually lint

For prettier: it adds trailing commas even though trailing commas are `none` in VSCode, which is due to the fact that running prettier on the command line is outside VSCode so the settings in `settings.json` is ignored and defaults is `all` (since version 3.0.0).

I can add a prettierrc.json to set it to `none` so it works but it was not included in the original project so I did neither. VSCode should format ok when save.

I am happy to make changes, I am not bias towards any kind of formatting (so when adding VSCode formatting settings I was trying to match the original as much as possible) so I would be happy to set whatever you would prefer. 

But I will not be able to fix 250+ eslint errors.

Let me know how to proceed: run prettier and add trailing commas as well as update VSCode formatting settings (my preference) or add prettier rule to disable trailingcommas.

PS: I have fixed a typo and removed an unnecessary async that I noticed

8beeeaaat commented:
@Abasz 

I'm sorry for the late reply! 
I've merged your kind comments and fixed the code formatting into the master branch. 
Please rebase or merge the latest master to your branch and run `npm run lint` .
If everything is fine, please review request with the push.

https://github.com/8beeeaaat/web-ant-plus/pull/7

Abasz commented:
> Please rebase or merge the latest master to your branch and run `npm run lint` .

Thanks for this. I will do the rebase and update the PR. Just a side remark:

You understand that this command `eslint --fix` does not do anything right? I mean that none of the eslint rules are enforced or run by this command. As I mentioned previously eslint cli needs the path to the file which it should lint. So the correct command would be `eslint --fix ./**/*.ts`, which would lint all `.ts` files. Again it only would if the .eslintrc.js file would be correct. But it is not as `"w3c-web-usb": true` is an invalid environment. So effectively you are not running eslint at all even if you run `npm run lint`

Abasz commented:
All done.

8beeeaaat reviewed (approved):
👍 👍 👍

8beeeaaat commented:
I updated the npm package!
Thanks for your contributing.

https://www.npmjs.com/package/web-ant-plus/v/1.0.5
